pallets/nfts/src/features/approvals.rs

```diff
@@ -165,10 +179,197 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
+	) -> DispatchResult {
+		ensure!(
+			Self::is_pallet_feature_enabled(PalletFeature::Approvals),
+			Error::<T, I>::MethodDisabled
+		);
+
+		let collection_config = Self::get_collection_config(&collection)?;
+		ensure!(
+			collection_config.is_setting_enabled(CollectionSetting::TransferableItems),
+			Error::<T, I>::ItemsNonTransferable
+		);
+		let now = frame_system::Pallet::<T>::block_number();
+		let deadline = maybe_deadline.map(|d| d.saturating_add(now));
+
+		CollectionApprovals::<T, I>::try_mutate_exists(
+			(&collection, &origin, &delegate),
+			|maybe_approval| -> DispatchResult {
+				let deposit_required = T::CollectionApprovalDeposit::get();
+				let mut current_deposit = match maybe_approval.take() {
+					Some((_, deposit)) => deposit,
+					None => Zero::zero(),
+				};
+
+				if current_deposit < deposit_required {
+					T::Currency::reserve(&origin, deposit_required - current_deposit)?;
+					current_deposit = deposit_required;
+				}
+				*maybe_approval = Some((deadline, current_deposit));
+				Ok(())
+			},
+		)?;
+
+		Self::deposit_event(Event::TransferApproved {
+			collection,
+			item: None,
+			owner: origin,
+			delegate,
+			deadline,
+		});
+
+		Ok(())
+	}
+
+	/// Cancels the transfer of items in the collection that owned by the origin to a delegate.
+	///
+	/// This function is used to cancel the approval for the transfer of items in the `collection`
+	/// that owned by the `origin` to a `delegate`. After canceling the approval, the function
+	/// returns the `origin` back the deposited fund and emits the `ApprovalCancelled` event.
+	///
+	/// - `origin`: The account grants permission to cancel the transfer.
+	/// - `collection`: The identifier of the collection.
+	/// - `delegate`: The account that was previously allowed to take control of items in the
+	///   collection that owned by the origin.
+	pub(crate) fn do_cancel_collection_approval(
+		origin: T::AccountId,
+		collection: T::CollectionId,
+		delegate: T::AccountId,
+	) -> DispatchResult {
+		let (_, deposit) = CollectionApprovals::<T, I>::take((&collection, &origin, &delegate))
+			.ok_or(Error::<T, I>::UnknownCollection)?;
+
+		T::Currency::unreserve(&origin, deposit);
+
+		Self::deposit_event(Event::ApprovalCancelled {
+			collection,
+			owner: origin,
+			item: None,
+			delegate,
+		});
+
+		Ok(())
+	}
+
+	/// Clears all collection approvals.
+	///
+	/// This function is used to clear all approvals to transfer items in the `collection` that
+	/// owned by the `origin` to a `delegate`. After clearing all approvals, the function returns
+	/// the `origin` back the deposited fund of each collection approval and emits the
+	/// `AllApprovalsCancelled` event.
+	///
+	/// - `origin`: The account grants permission to clear the transfer.
+	/// - `collection`: The collection ID containing the item.
+	/// - `witness_approvals`: Information on the collection approvals cleared. This must be
+	///   correct.
+	pub(crate) fn do_clear_all_collection_approvals(
```

Comment:
Perhaps this function should allow the Collection owner when all items are burned to execute. The owner of the approvals gets their funds back and the owner of the collection can remove it from storage in order to destroy the collection.

Reply:
shouldn't be the collection owner, this method is executed by any `origin`. I don't see any convincing reason why we block this method from being executed if there are still items. There are cases that I still want the collection items exist and just removing all approvals instead.

Reply:
I don't follow your last sentence, but oke, if it is allowed by anyone it makes sense, but only by anyone if all items are burned. Otherwise I can remove someone's approvals while the account has items